File: back-end-java/src/main/java/com/aie/backend/modules/govbr/service/GovbrService.java
package com.aie.backend.modules.govbr.service;

import java.net.URI;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.Base64;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Random;
import java.util.concurrent.ConcurrentHashMap;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.client.HttpStatusCodeException;
import org.springframework.stereotype.Service;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.server.ResponseStatusException;

import com.aie.backend.config.GovbrProperties;
import com.aie.backend.config.JwtProperties;
import com.aie.backend.modules.admin.repository.AdminRepository;
import com.aie.backend.modules.auth.dto.MenuItem;
import com.aie.backend.modules.auth.entities.Administrador;
import com.aie.backend.modules.auth.security.JwtTokenService;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.security.Keys;
import com.fasterxml.jackson.databind.ObjectMapper;

import org.springframework.http.HttpStatus;

@Service
public class GovbrService {

    private static final Logger LOGGER = LoggerFactory.getLogger(GovbrService.class);
    private final GovbrProperties properties;
    private final AdminRepository adminRepository;
    private final JwtTokenService tokenService;
    private final RestTemplate restTemplate = new RestTemplate();
    private final byte[] stateSecret;
    private final ObjectMapper objectMapper = new ObjectMapper();
    private final Map<String, PkceEntry> pkceStore = new ConcurrentHashMap<>();

    private final List<MenuItem> govbrMenu = List.of(
            new MenuItem("fas fa-chart-bar", "home", "sidebar.home", 0, ""),
            new MenuItem("fas fa-folder", "projects", "sidebar.projects", 1, "projects")
    );

    public GovbrService(GovbrProperties properties,
                        AdminRepository adminRepository,
                        JwtTokenService tokenService,
                        JwtProperties jwtProperties) {
        this.properties = properties;
        this.adminRepository = adminRepository;
        this.tokenService = tokenService;
        String secret = properties.getStateSecret() != null && !properties.getStateSecret().isBlank()
                ? properties.getStateSecret()
                : jwtProperties.getSecret();
        this.stateSecret = secret.getBytes(StandardCharsets.UTF_8);
    }

    public String buildAuthorizeUrl(String origin) {
        if (isBlank(properties.getClientId()) || isBlank(properties.getClientSecret())) {
            throw badRequest("Defina GOVBR_CLIENT_ID e GOVBR_CLIENT_SECRET para habilitar o login com gov.br");
        }

        String resolvedOrigin = resolveOrigin(origin);
        String codeVerifier = generateCodeVerifier();
        String codeChallenge = toCodeChallenge(codeVerifier);
        String state = signState(Map.of(
                "origin", resolvedOrigin,
                "ts", System.currentTimeMillis()
        ));
        storePkce(state, codeVerifier);

        String baseAuthUrl = notBlank(properties.getAuthUrl())
                ? properties.getAuthUrl()
                : String.format("%s/authorize", baseOrDefault(properties.getBaseUrl(), "https://sso.acesso.gov.br"));

        String nonce = randomHex(16);
        String params = "response_type=code" +
                "&client_id=" + url(properties.getClientId()) +
                "&redirect_uri=" + url(resolveRedirect()) +
                "&scope=" + url(properties.getScope()) +
                "&state=" + url(state) +
                "&nonce=" + url(nonce) +
                "&code_challenge=" + url(codeChallenge) +
                "&code_challenge_method=S256";

        return baseAuthUrl + "?" + params;
    }

    public Map<String, Object> handleCallback(String code, String state) {
        if (isBlank(code)) {
            throw badRequest("Código de autorização ausente no retorno do gov.br");
        }
        Map<String, Object> statePayload = verifyState(state);

        String codeVerifier = consumePkce(state);
        Map<String, Object> tokenSet = exchangeCodeForTokens(code, codeVerifier);
        Map<String, Object> profile = resolveProfile(tokenSet);
        Administrador user = findOrCreateUser(profile);

        Map<String, Object> payload = new HashMap<>();
        payload.put("id", user.getId());
        payload.put("name", user.getName());
        payload.put("email", user.getEmail());
        payload.put("social_number", user.getSocialNumber());
        payload.put("admin", Boolean.FALSE);
        payload.put("menu", govbrMenu);

        String token = tokenService.generateToken(payload);

        Map<String, Object> response = new HashMap<>();
        response.put("token", token);
        response.put("user", payload);
        response.put("origin", statePayload.getOrDefault("origin", resolveOrigin(null)));
        return response;
    }

    public String renderPopupClosePage(String origin, Map<String, Object> payload) {
        String target = origin != null ? origin : "*";
        String serialized = payloadToJson(payload);
        return "<!DOCTYPE html>\n" +
                "<html lang=\"pt-BR\">\n" +
                "<head><meta charset=\"utf-8\" /><title>gov.br</title></head>\n" +
                "<body>\n" +
                "  <script>\n" +
                "    (function() {\n" +
                "      const data = " + serialized + ";\n" +
                "      const targetOrigin = " + quote(target) + ";\n" +
                "      if (window.opener && !window.opener.closed) {\n" +
                "        try { window.opener.postMessage(data, targetOrigin === '*' ? '*' : targetOrigin); } catch (err) { console.error(err); }\n" +
                "      }\n" +
                "      window.close();\n" +
                "    })();\n" +
                "  </script>\n" +
                "  <p>Você pode fechar esta janela.</p>\n" +
                "</body>\n" +
                "</html>";
    }

    public String extractOriginFromState(String state) {
        try {
            Claims decoded = decodeToken(state);
            return decoded.get("origin", String.class);
        } catch (Exception ex) {
            LOGGER.warn("Não foi possível ler origin do state: {}", ex.getMessage());
            return null;
        }
    }

    private Map<String, Object> exchangeCodeForTokens(String code, String codeVerifier) {
        String tokenUrl = notBlank(properties.getTokenUrl())
                ? properties.getTokenUrl()
                : String.format("%s/token", baseOrDefault(properties.getBaseUrl(), "https://sso.acesso.gov.br"));

        MultiValueMap<String, String> body = new LinkedMultiValueMap<>();
        body.add("grant_type", "authorization_code");
        body.add("code", code);
        body.add("redirect_uri", resolveRedirect());
        body.add("client_id", properties.getClientId());
        body.add("client_secret", properties.getClientSecret());
        body.add("code_verifier", codeVerifier);

        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);
        headers.set(HttpHeaders.AUTHORIZATION, resolveBasicAuthorization());

        HttpEntity<MultiValueMap<String, String>> entity = new HttpEntity<>(body, headers);
        try {
            ResponseEntity<Map> response = restTemplate.postForEntity(URI.create(tokenUrl), entity, Map.class);
            Map<String, Object> data = response.getBody();
            if (data == null || isBlank((String) data.get("access_token"))) {
                throw unauthorized("Não recebemos access_token do gov.br");
            }
            return data;
        } catch (ResponseStatusException ex) {
            throw ex;
        } catch (HttpStatusCodeException ex) {
            LOGGER.error("Falha ao trocar code por token gov.br status={} body={}",
                    ex.getStatusCode().value(),
                    sanitizeTokenErrorBody(ex.getResponseBodyAsString()));
            throw unauthorized("Falha ao autenticar no gov.br");
        } catch (Exception ex) {
            LOGGER.error("Falha ao trocar code por token gov.br", ex);
            throw unauthorized("Falha ao autenticar no gov.br");
        }
    }

    private Map<String, Object> resolveProfile(Map<String, Object> tokenSet) {
        Map<String, Object> decoded = decodeIdToken((String) tokenSet.get("id_token"));
        Map<String, Object> userinfo = fetchUserinfo((String) tokenSet.get("access_token"));

        Map<String, Object> data = new HashMap<>();
        if (userinfo != null) data.putAll(userinfo);
        if (decoded != null) data.putAll(decoded);

        String name = optionalString(data, "name", "nome", "given_name");
        if (isBlank(name)) name = "Usuário gov.br";

        String email = optionalString(data, "email", "preferred_username");
        String social = formatCpf(optionalString(data, "cpf", "social_number", "sub"));

        Map<String, Object> profile = new HashMap<>();
        profile.put("name", name);
        profile.put("email", email);
        profile.put("social_number", social);
        profile.put("raw", Map.of("id_token", decoded, "userinfo", userinfo));
        return profile;
    }

    private Administrador findOrCreateUser(Map<String, Object> profile) {
        String social = (String) profile.get("social_number");
        String email = (String) profile.get("email");

        Optional<Administrador> existing = Optional.empty();
        if (!isBlank(email) || !isBlank(social)) {
            existing = adminRepository.findByEmailIgnoreCaseOrSocialNumber(
                    email != null ? email : "",
                    social != null ? social : "");
        }

        if (existing.isPresent()) {
            Administrador admin = existing.get();
            admin.setName((String) profile.getOrDefault("name", admin.getName()));
            if (!isBlank(email)) admin.setEmail(email);
            if (!isBlank(social)) admin.setSocialNumber(social);
            admin.setLastAccess(Instant.now());
            return adminRepository.save(admin);
        }

        Administrador admin = new Administrador();
        admin.setName((String) profile.getOrDefault("name", "Usuário gov.br"));
        admin.setEmail(!isBlank(email) ? email : placeholderEmail());
        admin.setSocialNumber(!isBlank(social) ? social : generateFallbackSocial());
        admin.setPosition("gov.br");
        admin.setActive(Boolean.TRUE);
        admin.setPassword(null);
        admin.setLastAccess(Instant.now());
        return adminRepository.save(admin);
    }

    private Map<String, Object> fetchUserinfo(String accessToken) {
        if (isBlank(accessToken)) return null;

        String url = notBlank(properties.getUserinfoUrl())
                ? properties.getUserinfoUrl()
                : String.format("%s/userinfo", baseOrDefault(properties.getApiBaseUrl(), "https://api.acesso.gov.br"));

        HttpHeaders headers = new HttpHeaders();
        headers.set(HttpHeaders.AUTHORIZATION, "Bearer " + accessToken);
        HttpEntity<Void> entity = new HttpEntity<>(headers);
        try {
            ResponseEntity<Map> response = restTemplate.exchange(URI.create(url), HttpMethod.GET, entity, Map.class);
            return response.getBody();
        } catch (Exception ex) {
            LOGGER.warn("Não foi possível obter userinfo gov.br: {}", ex.getMessage());
            return null;
        }
    }

    private Map<String, Object> decodeIdToken(String idToken) {
        if (isBlank(idToken)) return null;
        try {
            String[] parts = idToken.split("\\.");
            if (parts.length < 2) return null;
            String payloadJson = new String(Base64.getUrlDecoder().decode(parts[1]), StandardCharsets.UTF_8);
            return objectMapper.readValue(payloadJson, Map.class);
        } catch (Exception ex) {
            LOGGER.warn("Falha ao decodificar id_token gov.br: {}", ex.getMessage());
            return null;
        }
    }

    private String signState(Map<String, Object> payload) {
        Instant now = Instant.now();
        Instant exp = now.plusSeconds(600);
        return Jwts.builder()
                .setClaims(payload)
                .setIssuedAt(java.util.Date.from(now))
                .setExpiration(java.util.Date.from(exp))
                .signWith(Keys.hmacShaKeyFor(stateSecret))
                .compact();
    }

    private Map<String, Object> verifyState(String state) {
        if (isBlank(state)) throw badRequest("State não recebido do gov.br");
        try {
            Claims claims = Jwts.parserBuilder()
                    .setSigningKey(Keys.hmacShaKeyFor(stateSecret))
                    .build()
                    .parseClaimsJws(state)
                    .getBody();
            Map<String, Object> map = new HashMap<>();
            claims.forEach(map::put);
            return map;
        } catch (Exception ex) {
            LOGGER.warn("State inválido ou expirado: {}", ex.getMessage());
            throw badRequest("State inválido ou expirado");
        }
    }

    private String resolveRedirect() {
        return notBlank(properties.getRedirectUri())
                ? properties.getRedirectUri()
                : "http://localhost:8080/retornoWebHook";
    }

    private String resolveOrigin(String origin) {
        if (notBlank(origin)) return origin;
        if (notBlank(properties.getFrontendOrigin())) return properties.getFrontendOrigin();
        return "http://localhost:3000";
    }

    private String payloadToJson(Map<String, Object> payload) {
        Map<String, Object> enriched = new HashMap<>(payload);
        enriched.put("source", "govbr-login");
        try {
            return objectMapper.writeValueAsString(enriched);
        } catch (Exception ex) {
            return "{\"status\":\"error\",\"message\":\"Falha ao serializar resposta\"}";
        }
    }

    private String optionalString(Map<String, Object> map, String... keys) {
        for (String key : keys) {
            Object val = map.get(key);
            if (val != null && !String.valueOf(val).isBlank()) {
                return String.valueOf(val);
            }
        }
        return null;
    }

    private String formatCpf(String value) {
        if (isBlank(value)) return null;
        String digits = value.replaceAll("\\D", "");
        if (digits.length() < 11) return digits;
        String trimmed = digits.substring(0, 11);
        return trimmed.replaceAll("(\\d{3})(\\d{3})(\\d{3})(\\d{2})", "$1.$2.$3-$4");
    }

    private String generateFallbackSocial() {
        Random random = new Random();
        StringBuilder digits = new StringBuilder();
        for (int i = 0; i < 11; i++) digits.append(random.nextInt(10));
        return formatCpf(digits.toString());
    }

    private String placeholderEmail() {
        return "govbr-" + Long.toHexString(System.currentTimeMillis()) + "@placeholder.local";
    }

    private String url(String value) {
        return URLEncoder.encode(value, StandardCharsets.UTF_8);
    }

    private boolean isBlank(String value) {
        return value == null || value.trim().isEmpty();
    }

    private boolean notBlank(String value) {
        return !isBlank(value);
    }

    private String baseOrDefault(String value, String fallback) {
        return notBlank(value) ? value : fallback;
    }

    private String randomHex(int bytes) {
        byte[] buffer = new byte[bytes];
        new Random().nextBytes(buffer);
        return Base64.getUrlEncoder().withoutPadding().encodeToString(buffer);
    }

    private void storePkce(String state, String codeVerifier) {
        cleanupExpiredPkce();
        pkceStore.put(state, new PkceEntry(codeVerifier, System.currentTimeMillis() + 10 * 60 * 1000));
    }

    private String consumePkce(String state) {
        if (isBlank(state)) {
            throw badRequest("State não recebido do gov.br");
        }
        cleanupExpiredPkce();
        PkceEntry entry = pkceStore.remove(state);
        if (entry == null || entry.expiresAt() < System.currentTimeMillis()) {
            throw badRequest("Sessão PKCE não encontrada ou expirada. Inicie o login novamente.");
        }
        return entry.codeVerifier();
    }

    private void cleanupExpiredPkce() {
        long now = System.currentTimeMillis();
        pkceStore.entrySet().removeIf(e -> e.getValue().expiresAt() < now);
    }

    private String generateCodeVerifier() {
        byte[] bytes = new byte[32];
        new Random().nextBytes(bytes);
        return base64Url(bytes);
    }

    private String toCodeChallenge(String codeVerifier) {
        try {
            java.security.MessageDigest digest = java.security.MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(codeVerifier.getBytes(StandardCharsets.UTF_8));
            return base64Url(hash);
        } catch (Exception ex) {
            throw new IllegalStateException("Não foi possível gerar code_challenge PKCE", ex);
        }
    }

    private String base64Url(byte[] value) {
        return Base64.getUrlEncoder().withoutPadding().encodeToString(value);
    }

    private String resolveBasicAuthorization() {
        if (notBlank(properties.getBasicAuth())) {
            String basic = properties.getBasicAuth();
            return basic.startsWith("Basic ") ? basic : "Basic " + basic;
        }
        String raw = properties.getClientId() + ":" + properties.getClientSecret();
        return "Basic " + Base64.getEncoder().encodeToString(raw.getBytes(StandardCharsets.UTF_8));
    }

    private String sanitizeTokenErrorBody(String body) {
        if (body == null) return null;
        if (body.length() > 1000) {
            return body.substring(0, 1000);
        }
        return body;
    }

    private ResponseStatusException badRequest(String msg) {
        return new ResponseStatusException(HttpStatus.BAD_REQUEST, msg);
    }

    private ResponseStatusException unauthorized(String msg) {
        return new ResponseStatusException(HttpStatus.UNAUTHORIZED, msg);
    }

    private Claims decodeToken(String token) {
        return Jwts.parserBuilder()
                .setSigningKey(Keys.hmacShaKeyFor(stateSecret))
                .build()
                .parseClaimsJws(token)
                .getBody();
    }

    private String quote(String value) {
        return "\"" + escape(value) + "\"";
    }

    private String escape(String value) {
        return value.replace("\\", "\\\\").replace("\"", "\\\"");
    }

    private record PkceEntry(String codeVerifier, long expiresAt) {}
}

File: back-end-nestjs/src/modules/govbr/govbr.service.ts
/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */

import {
  BadRequestException,
  Injectable,
  Logger,
  UnauthorizedException,
} from '@nestjs/common';
import { HttpService } from '@nestjs/axios';
import { firstValueFrom } from 'rxjs';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import * as crypto from 'crypto';
import * as jwt from 'jsonwebtoken';
import { JwtService } from '@nestjs/jwt';
import { Administrador } from '../auth/entities/administrador.entity';

type GovbrTokenResponse = {
  access_token: string;
  id_token?: string;
  token_type?: string;
  expires_in?: number;
  refresh_token?: string;
  scope?: string;
};

type GovbrProfile = {
  name?: string;
  email?: string | null;
  social_number?: string | null;
  raw?: any;
};

type PopupPayload =
  | { status: 'success'; token: string; user: any }
  | { status: 'error'; message: string; error?: string };

@Injectable()
export class GovbrService {
  private readonly logger = new Logger(GovbrService.name);
  private readonly pkceStore = new Map<
    string,
    { codeVerifier: string; expiresAt: number }
  >();
  private readonly baseAuthUrl =
    process.env.GOVBR_AUTH_URL ||
    `${process.env.GOVBR_BASE_URL || 'https://sso.acesso.gov.br'}/authorize`;
  private readonly tokenUrl =
    process.env.GOVBR_TOKEN_URL ||
    `${process.env.GOVBR_BASE_URL || 'https://sso.acesso.gov.br'}/token`;
  private readonly userinfoUrl =
    process.env.GOVBR_USERINFO_URL ||
    `${process.env.GOVBR_API_BASE_URL || 'https://api.acesso.gov.br'}/userinfo`;
  private readonly redirectUri =
    process.env.GOVBR_REDIRECT_URI || 'http://localhost:8080/retornoWebHook';
  private readonly frontendOrigin =
    process.env.GOVBR_FRONTEND_ORIGIN || 'http://localhost:3000';
  private readonly scope =
    process.env.GOVBR_SCOPE || 'openid email profile';
  private readonly stateSecret =
    process.env.GOVBR_STATE_SECRET ||
    process.env.JWT_SECRET ||
    'govbr-state-secret';

  private readonly govbrMenu = [
    {
      icon: 'fas fa-chart-bar',
      id: 'home',
      name: 'sidebar.home',
      order: 0,
      path: '',
    },
    {
      icon: 'fas fa-folder',
      id: 'projects',
      name: 'sidebar.projects',
      order: 1,
      path: 'projects',
    },
  ];

  constructor(
    private readonly http: HttpService,
    @InjectRepository(Administrador)
    private readonly adminRepo: Repository<Administrador>,
    private readonly jwtService: JwtService,
  ) {}

  async buildAuthorizeUrl(origin?: string) {
    const clientId = process.env.GOVBR_CLIENT_ID;
    const clientSecret = process.env.GOVBR_CLIENT_SECRET;
    if (!clientId || !clientSecret) {
      throw new BadRequestException(
        'Defina GOVBR_CLIENT_ID e GOVBR_CLIENT_SECRET para habilitar o login com gov.br',
      );
    }

    const codeVerifier = this.generateCodeVerifier();
    const codeChallenge = this.toCodeChallenge(codeVerifier);

    const state = await this.jwtService.signAsync(
      {
        origin: this.resolveOrigin(origin),
        ts: Date.now(),
      },
      {
        secret: this.stateSecret,
        expiresIn: '10m',
      },
    );
    this.storePkce(state, codeVerifier);

    const params = new URLSearchParams({
      response_type: 'code',
      client_id: clientId,
      redirect_uri: this.redirectUri,
      scope: this.scope,
      state,
      nonce: crypto.randomBytes(16).toString('hex'),
      code_challenge: codeChallenge,
      code_challenge_method: 'S256',
    });

    return `${this.baseAuthUrl}?${params.toString()}`;
  }

  async handleCallback(code?: string, state?: string) {
    if (!code) {
      throw new BadRequestException(
        'Código de autorização ausente no retorno do gov.br',
      );
    }

    const statePayload = await this.verifyState(state);
    const codeVerifier = this.consumePkce(state);
    const tokenSet = await this.exchangeCodeForTokens(code, codeVerifier);
    const profile = await this.resolveProfile(tokenSet);
    const userEntity = await this.findOrCreateUser(profile);

    await this.adminRepo.update(userEntity.id, { last_access: new Date() });

    const userPayload = {
      id: userEntity.id,
      name: userEntity.name,
      email: userEntity.email,
      social_number: userEntity.social_number,
      admin: false,
      menu: this.govbrMenu,
    };

    const token = await this.jwtService.signAsync(userPayload);

    return {
      token,
      user: userPayload,
      origin: statePayload?.origin ?? this.frontendOrigin,
    };
  }

  renderPopupClosePage(origin: string, payload: PopupPayload) {
    const targetOrigin = origin || '*';
    const serialized = JSON.stringify({
      source: 'govbr-login',
      ...payload,
    });

    return `<!DOCTYPE html>
<html lang="pt-BR">
<head>
  <meta charset="utf-8" />
  <title>gov.br</title>
</head>
<body>
  <script>
    (function() {
      const data = ${serialized};
      const targetOrigin = ${JSON.stringify(targetOrigin)};
      if (window.opener && !window.opener.closed) {
        try { window.opener.postMessage(data, targetOrigin === '*' ? '*' : targetOrigin); } catch (err) { console.error(err); }
      }
      window.close();
    })();
  </script>
  <p>Você pode fechar esta janela.</p>
</body>
</html>`;
  }

  extractOriginFromState(state?: string) {
    if (!state) return null;
    try {
      const decoded: any = jwt.decode(state);
      if (typeof decoded === 'object' && decoded?.origin)
        return decoded.origin as string;
    } catch (err) {
      this.logger.warn('Não foi possível ler o origin do state recebido.');
    }
    return null;
  }

  getOriginForResponse(
    originFromQuery?: string,
    originFromState?: string | null,
  ) {
    return this.resolveOrigin(originFromQuery || originFromState || undefined);
  }

  private resolveOrigin(origin?: string) {
    return origin || this.frontendOrigin;
  }

  private async verifyState(
    state?: string,
  ): Promise<{ origin?: string } | null> {
    if (!state) {
      throw new BadRequestException('State não recebido do gov.br');
    }
    try {
      return await this.jwtService.verifyAsync(state, {
        secret: this.stateSecret,
      });
    } catch (err) {
      this.logger.warn('State inválido ou expirado no retorno do gov.br');
      throw new BadRequestException('State inválido ou expirado');
    }
  }

  private async exchangeCodeForTokens(
    code: string,
    codeVerifier: string,
  ): Promise<GovbrTokenResponse> {
    const clientId = process.env.GOVBR_CLIENT_ID;
    const clientSecret = process.env.GOVBR_CLIENT_SECRET;
    const basicAuth = process.env.GOVBR_BASIC_AUTH;

    if (!clientId || !clientSecret) {
      throw new BadRequestException(
        'Credenciais do gov.br não configuradas (GOVBR_CLIENT_ID / GOVBR_CLIENT_SECRET).',
      );
    }

    const body = new URLSearchParams({
      grant_type: 'authorization_code',
      code,
      redirect_uri: this.redirectUri,
      client_id: clientId,
      client_secret: clientSecret,
      code_verifier: codeVerifier,
    });

    try {
      const headers: Record<string, string> = {
        'Content-Type': 'application/x-www-form-urlencoded',
      };

      headers.Authorization = this.resolveBasicAuthorization(
        clientId,
        clientSecret,
        basicAuth,
      );

      const response = await firstValueFrom(
        this.http.post<GovbrTokenResponse>(this.tokenUrl, body.toString(), {
          headers,
        }),
      );
      if (!response.data?.access_token) {
        throw new UnauthorizedException(
          'Não recebemos o access_token do gov.br',
        );
      }
      return response.data;
    } catch (err: any) {
      const status = err?.response?.status ?? null;
      const data = this.sanitizeTokenErrorResponse(err?.response?.data);
      this.logger.error(
        `Falha ao trocar o code pelo token do gov.br (status=${status ?? 'n/a'}) body=${JSON.stringify(data)}`,
      );
      throw new UnauthorizedException('Falha ao autenticar no gov.br');
    }
  }

  private async resolveProfile(
    tokenSet: GovbrTokenResponse,
  ): Promise<GovbrProfile> {
    const decoded = tokenSet.id_token
      ? ((jwt.decode(tokenSet.id_token) as any) ?? {})
      : {};
    let userInfo: any = null;

    if (tokenSet.access_token) {
      try {
        const response = await firstValueFrom(
          this.http.get(this.userinfoUrl, {
            headers: { Authorization: `Bearer ${tokenSet.access_token}` },
          }),
        );
        userInfo = response.data;
      } catch (err) {
        this.logger.warn(
          'Não foi possível obter o userinfo do gov.br, tentando seguir com o id_token.',
        );
      }
    }

    const data = {
      ...(userInfo ?? {}),
      ...(decoded ?? {}),
    };

    const name =
      data?.name || data?.nome || data?.given_name || 'Usuário gov.br';
    const email = data?.email || data?.preferred_username || null;
    const socialNumber =
      this.formatCpf(data?.cpf || data?.social_number || data?.sub) ?? null;

    return {
      name,
      email,
      social_number: socialNumber,
      raw: { id_token: decoded, userinfo: userInfo },
    };
  }

  private async findOrCreateUser(
    profile: GovbrProfile,
  ): Promise<Administrador> {
    const qb = this.adminRepo.createQueryBuilder('adm');

    if (profile.social_number && profile.email) {
      qb.where('adm.social_number = :social_number OR adm.email = :email', {
        social_number: profile.social_number,
        email: profile.email,
      });
    } else if (profile.social_number) {
      qb.where('adm.social_number = :social_number', {
        social_number: profile.social_number,
      });
    } else if (profile.email) {
      qb.where('adm.email = :email', { email: profile.email });
    }

    const existing = await qb.getOne();

    if (existing) {
      await this.adminRepo.update(existing.id, {
        name: profile.name ?? existing.name,
        email: profile.email ?? existing.email,
        social_number: profile.social_number ?? existing.social_number,
      });
      const refreshed = await this.adminRepo.findOne({
        where: { id: existing.id },
      });
      return refreshed ?? existing;
    }

    const social_number =
      profile.social_number ?? this.generateFallbackSocialNumber();

    const newUser = this.adminRepo.create({
      name: profile.name ?? 'Usuário gov.br',
      email:
        profile.email ??
        `govbr-${Math.random().toString(36).substring(2, 8)}@placeholder.local`,
      social_number,
      position: 'gov.br',
      active: true,
      last_access: new Date(),
    });

    return this.adminRepo.save(newUser);
  }

  private formatCpf(value?: string | null) {
    if (!value) return null;
    const digits = value.replace(/\D/g, '');
    const trimmed = digits.slice(0, 14);
    if (digits.length !== 11) return trimmed || null;
    return digits.replace(
      /(\d{3})(\d{3})(\d{3})(\d{2})/,
      (_, p1, p2, p3, p4) => `${p1}.${p2}.${p3}-${p4}`,
    );
  }

  private generateFallbackSocialNumber() {
    const random = Math.floor(1e10 + Math.random() * 9e10)
      .toString()
      .slice(0, 11);
    return this.formatCpf(random) ?? random.padEnd(11, '0').slice(0, 11);
  }

  private storePkce(state: string, codeVerifier: string) {
    this.cleanupExpiredPkce();
    this.pkceStore.set(state, {
      codeVerifier,
      expiresAt: Date.now() + 10 * 60 * 1000,
    });
  }

  private consumePkce(state?: string) {
    if (!state) {
      throw new BadRequestException('State não recebido do gov.br');
    }
    this.cleanupExpiredPkce();
    const entry = this.pkceStore.get(state);
    if (!entry || entry.expiresAt < Date.now()) {
      this.pkceStore.delete(state);
      throw new BadRequestException(
        'Sessão PKCE não encontrada ou expirada. Inicie o login novamente.',
      );
    }
    this.pkceStore.delete(state);
    return entry.codeVerifier;
  }

  private cleanupExpiredPkce() {
    const now = Date.now();
    for (const [key, value] of this.pkceStore.entries()) {
      if (value.expiresAt < now) {
        this.pkceStore.delete(key);
      }
    }
  }

  private generateCodeVerifier() {
    return this.base64UrlEncode(crypto.randomBytes(32));
  }

  private toCodeChallenge(codeVerifier: string) {
    const hash = crypto.createHash('sha256').update(codeVerifier).digest();
    return this.base64UrlEncode(hash);
  }

  private base64UrlEncode(value: Buffer) {
    return value
      .toString('base64')
      .replace(/\+/g, '-')
      .replace(/\//g, '_')
      .replace(/=+$/g, '');
  }

  private resolveBasicAuthorization(
    clientId: string,
    clientSecret: string,
    basicAuth?: string,
  ) {
    if (basicAuth) {
      return basicAuth.startsWith('Basic ') ? basicAuth : `Basic ${basicAuth}`;
    }

    const raw = Buffer.from(`${clientId}:${clientSecret}`).toString('base64');
    return `Basic ${raw}`;
  }

  private sanitizeTokenErrorResponse(data: unknown) {
    if (!data || typeof data !== 'object') return data;
    const clone = { ...(data as Record<string, unknown>) };
    if ('access_token' in clone) clone.access_token = '[redacted]';
    if ('refresh_token' in clone) clone.refresh_token = '[redacted]';
    if ('id_token' in clone) clone.id_token = '[redacted]';
    if ('client_secret' in clone) clone.client_secret = '[redacted]';
    return clone;
  }
}
